c/rce.h

```diff
@@ -0,0 +1,382 @@
+                 size_t args_len) {
+  int err = 0;
+  RceState rce_state;
+  rce_init_state(&rce_state);
+
+  uint32_t index = 0;
+
+  CHECK2(args_len == BLAKE2B_BLOCK_SIZE, ERROR_INVALID_MOL_FORMAT);
+  CHECK2(args != NULL, ERROR_INVALID_RCE_ARGS);
+  if (is_owner_mode) return 0;
+
+  rce_state.rcrules_count = 0;
+  err = rce_gather_rcrules_recursively(&rce_state, args, 0);
+  CHECK(err);
+
+  SmtProofVecType proofs;
+  err = rce_get_proofs(extension_index, &proofs);
+  CHECK(err);
+
+  uint32_t proof_len = proofs.t->len(&proofs);
+  // count of proof should be same as size of RCRules
+  CHECK2(proof_len == rce_state.rcrules_count, ERROR_RCRULES_PROOFS_MISMATCHED);
+
+  smt_pair_t wl_entries[MAX_LOCK_SCRIPT_HASH_COUNT];
+  smt_pair_t bl_entries[MAX_LOCK_SCRIPT_HASH_COUNT];
+  smt_state_t wl_states;
+  smt_state_t bl_states;
+  smt_state_init(&wl_states, wl_entries, MAX_LOCK_SCRIPT_HASH_COUNT);
+  smt_state_init(&bl_states, bl_entries, MAX_LOCK_SCRIPT_HASH_COUNT);
+
+  err = rce_collect_hashes(&bl_states, &wl_states);
+  CHECK(err);
+
+  smt_state_normalize(&wl_states);
+  smt_state_normalize(&bl_states);
+
+  err = ERROR_SMT_VERIFY_FAILED;
+  bool on_white_list = false;
+  bool on_black_list = false;
+
+  uint8_t temp_proof[MAX_TEMP_PROOF_LENGTH];
+
+  for (index = 0; index < proof_len; index++) {
+    bool existing = false;
+    mol2_cursor_t proof = proofs.t->get(&proofs, index, &existing);
+    CHECK2(existing, ERROR_INVALID_MOL_FORMAT);
+
+    const RCRule* current_rule = &rce_state.rcrules[index];
+
+    const uint8_t* root_hash = current_rule->smt_root;
+    // "Current RCRule must not be in Emergency Halt mode."
+    if (rce_is_emergency_halt_mode(current_rule->flags)) {
+      err = ERROR_RCE_EMERGENCY_HALT;
+      // the emergency halt has the highest priority, can return immediately
+      goto exit;
+    }
+    uint32_t temp_proof_len =
+        mol2_read_at(&proof, temp_proof, MAX_TEMP_PROOF_LENGTH);
+    CHECK2(temp_proof_len <= MAX_TEMP_PROOF_LENGTH, ERROR_TOO_LONG_PROOF);
+
+    if (rce_is_white_list(current_rule->flags)) {
+      err = smt_verify(root_hash, &wl_states, temp_proof, temp_proof_len);
+      // For all RCRules using whitelists, as long as there is one RCRule which
+      // satisfies err == 0, we consider validation to be success.
+      if (err == 0) {
+        // *** all hashes on white list
+        on_white_list = true;
+        // don't break when verify successfully: it's possible to be on
+        // emergency halt afterward.
+      }
+      //  else *** not all hashes on white list
+    } else {
+      // black list
+      err = smt_verify(root_hash, &bl_states, temp_proof, temp_proof_len);
+      if (err != 0) {
+        // *** any one of hashes on black list
+        on_black_list = true;
+      }
+      //  else *** all hashes *not* on black list
+
+      // don't break when verify failed: it's possible to be on white list
+      // afterward.
+    }
+  }
```

Comment:
We can sort `RceState.rcrules` and `proofs` their priority (`RCRule.flags`):
1. EMERGENCY_HALT
2. white list
3. black list

Then we can early return the rule verification.

Reply:
It's a good idea if we meet cycles bottleneck. Not optimize it too early.